crates/sui-framework/deps/move-stdlib/sources/u64.move

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Basic math for nicer programmability
+module std::u64 {
+    use std::ascii;
+    use std::option::{Self, Option};
+    use std::string;
+    use std::vector;
+
+    /// Largest possible u64 value
+    const MAX: u64 = 0;
+
+    /// Attempting to perform an invalid operation (e.g., mean) on an empty vector
+    const EVectorEmpty: u64 = 0;
+
+    /// Return the larger of `x` and `y`
+    public fun max(x: u64, y: u64): u64 {
+        if (x > y) {
+            x
+        } else {
+            y
+        }
+    }
+
+    /// Return the smaller of `x` and `y`
+    public fun min(x: u64, y: u64): u64 {
+        if (x < y) {
+            x
+        } else {
+            y
+        }
+    }
+
+    /// Return the value of a base raised to a power.
+    /// This function is not overflow-safe and will abort with an overflow error if `base^exponent` > `MAX`
+    public fun pow(base: u64, exponent: u8): u64 {
+        let res = 1;
+        while (exponent >= 1) {
+            if (exponent % 2 == 0) {
+                base = base * base;
+                exponent = exponent / 2;
+            } else {
+                res = res * base;
+                exponent = exponent - 1;
+            }
+        };
+
+        res
+    }
+
+    /// Return the mean of x and y, with the remainder truncated.
+    /// This function is overflow-safe and cannot abort
+    public fun mean(x: u64, y: u64): u64 {
+        if (x < y) {
+            x + (y - x) / 2
+        } else {
+            y + (x - y) / 2
+        }
+    }
+
+    /// Get a nearest lower integer square root for `x`. Given that this
+    /// function can only operate with integers, it is impossible
+    /// to get perfect (or precise) integer square root for some numbers.
+    ///
+    /// Example:
+    /// ```
+    /// math::sqrt(9) => 3
+    /// math::sqrt(8) => 2 // the nearest lower square root is 4;
+    /// ```
+    ///
+    /// In integer math, one of the possible ways to get results with more
+    /// precision is to use higher values or temporarily multiply the
+    /// value by some bigger number. Ideally if this is a square of 10 or 100.
+    ///
+    /// Example:
+    /// ```
+    /// math::sqrt(8) => 2;
+    /// math::sqrt(8 * 10000) => 282;
+    /// // now we can use this value as if it was 2.82;
+    /// // but to get the actual result, this value needs
+    /// // to be divided by 100 (because sqrt(10000)).
+    ///
+    ///
+    /// math::sqrt(8 * 1000000) => 2828; // same as above, 2828 / 1000 (2.828)
+    /// ```
+    public fun sqrt(x: u64): u64 {
```

Comment:
@sblackshear I was intrigued, so I let #chatGPT to contribute to this review and compare against your scheme :)
<img width="610" alt="Screen Shot 2022-12-15 at 9 27 30 PM" src="https://user-images.githubusercontent.com/25659162/208028856-299e5f54-9d9b-495f-859c-dcba4916a378.png">